# Nama: Muhammad Kaisar Novrenza
# NIM: 2602113546

## Import Library

In [1]:
import pandas as pd

from sklearn.preprocessing import StandardScaler
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import OneHotEncoder

import pickle

from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.tree import DecisionTreeClassifier

from sklearn.model_selection import train_test_split
from sklearn.pipeline import Pipeline
from sklearn.metrics import accuracy_score, classification_report

import streamlit as st
import joblib

# Load Dataset

In [2]:
df = pd.read_csv('B.csv')
print(df.head())

   student_id  gender  ssc_percentage  hsc_percentage  degree_percentage  \
0           1    Male              53              79                 56   
1           2  Female              56              54                 59   
2           3    Male              94              83                 83   
3           4    Male              84              71                 87   
4           5    Male              58              88                 74   

   cgpa  entrance_exam_score  technical_skill_score  soft_skill_score  \
0  8.87                   50                     92                90   
1  6.78                   61                     51                99   
2  7.92                   91                     93                84   
3  6.57                   85                     60                72   
4  9.01                   73                     52                88   

   internship_count  live_projects  work_experience_months  certifications  \
0                 1       

## Checking Null/Missing Values

In [3]:
missing_values = df.isnull().sum()
print(missing_values)

student_id                    0
gender                        0
ssc_percentage                0
hsc_percentage                0
degree_percentage             0
cgpa                          0
entrance_exam_score           0
technical_skill_score         0
soft_skill_score              0
internship_count              0
live_projects                 0
work_experience_months        0
certifications                0
attendance_percentage         0
backlogs                      0
extracurricular_activities    0
placement_status              0
salary_package_lpa            0
dtype: int64


## Correction Analysis

In [4]:
numeric_data = df.select_dtypes(include=['int64', 'float64'])
correlation_matrix = numeric_data.corr()
print(correlation_matrix)

                        student_id  ssc_percentage  hsc_percentage  \
student_id                1.000000       -0.004994        0.001606   
ssc_percentage           -0.004994        1.000000       -0.019349   
hsc_percentage            0.001606       -0.019349        1.000000   
degree_percentage        -0.001868        0.009593       -0.012460   
cgpa                      0.016297        0.011572       -0.012269   
entrance_exam_score      -0.007966       -0.016133       -0.010286   
technical_skill_score     0.008262       -0.017894        0.004797   
soft_skill_score         -0.007278       -0.012745       -0.002642   
internship_count         -0.002550       -0.022824       -0.012104   
live_projects            -0.022502        0.008257       -0.019474   
work_experience_months   -0.015962        0.009549       -0.014230   
certifications           -0.000635       -0.021682       -0.024707   
attendance_percentage     0.006192       -0.013790       -0.001179   
backlogs            

# Feature Engineering

## One hot encoding

In [5]:
data_encoded = pd.get_dummies(df, columns=['gender'],drop_first=True)
data_encoded['extracurricular_activities'] = data_encoded['extracurricular_activities'].map({'Yes': 1, 'No': 0})
print(data_encoded.head())

   student_id  ssc_percentage  hsc_percentage  degree_percentage  cgpa  \
0           1              53              79                 56  8.87   
1           2              56              54                 59  6.78   
2           3              94              83                 83  7.92   
3           4              84              71                 87  6.57   
4           5              58              88                 74  9.01   

   entrance_exam_score  technical_skill_score  soft_skill_score  \
0                   50                     92                90   
1                   61                     51                99   
2                   91                     93                84   
3                   85                     60                72   
4                   73                     52                88   

   internship_count  live_projects  work_experience_months  certifications  \
0                 1              3                      23               4

## Train Test Split

In [6]:
X = data_encoded.drop(columns=['student_id', 'placement_status', 'salary_package_lpa'])
y = data_encoded['placement_status']
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

print(X_train.shape, X_test.shape, y_train.shape, y_test.shape)

(4000, 15) (1000, 15) (4000,) (1000,)


## Logisitic Regression Model

In [7]:
logreg_model = LogisticRegression(max_iter=1000, random_state=42)
logreg_model.fit(X_train, y_train)
y_pred_logreg = logreg_model.predict(X_test)

log_accuracy = accuracy_score(y_test, y_pred_logreg)
log_classification_report = classification_report(y_test, y_pred_logreg)

print(f"Logistic Regression Accuracy: {log_accuracy}")
print("Classification Report for Logistic Regression:")
print(log_classification_report)

Logistic Regression Accuracy: 0.876
Classification Report for Logistic Regression:
              precision    recall  f1-score   support

           0       0.90      0.95      0.92       801
           1       0.75      0.57      0.65       199

    accuracy                           0.88      1000
   macro avg       0.82      0.76      0.79      1000
weighted avg       0.87      0.88      0.87      1000



Saya memilih menggunakan logistic regression karena untuk data tipe binary classification itu simple dan sangat efektif yang merupakan efesien untuk data yang memiliki hubungan linear antara features dan target variable.

## Random Forest Model

In [8]:
rf_model = RandomForestClassifier(random_state=42)
rf_model.fit(X_train, y_train)
y_pred_rf = rf_model.predict(X_test)

rf_accuracy = accuracy_score(y_test, y_pred_rf)
rf_classification_report = classification_report(y_test, y_pred_rf)

print(f"Random Forest Accuracy: {rf_accuracy}")
print("Classification Report for Random Forest:")
print(rf_classification_report)

Random Forest Accuracy: 1.0
Classification Report for Random Forest:
              precision    recall  f1-score   support

           0       1.00      1.00      1.00       801
           1       1.00      1.00      1.00       199

    accuracy                           1.00      1000
   macro avg       1.00      1.00      1.00      1000
weighted avg       1.00      1.00      1.00      1000



## Decision Tree Model

In [9]:
dt_model = DecisionTreeClassifier(random_state=42)
dt_model.fit(X_train, y_train)
y_pred_dt = dt_model.predict(X_test)

dt_accuracy = accuracy_score(y_test, y_pred_dt)
dt_classification_report = classification_report(y_test, y_pred_dt)

print(f"Decision Tree Accuracy: {dt_accuracy}")
print("Classification Report for Decision Tree:")
print(dt_classification_report)

Decision Tree Accuracy: 1.0
Classification Report for Decision Tree:
              precision    recall  f1-score   support

           0       1.00      1.00      1.00       801
           1       1.00      1.00      1.00       199

    accuracy                           1.00      1000
   macro avg       1.00      1.00      1.00      1000
weighted avg       1.00      1.00      1.00      1000



Penjelasan

Evaluation Metrics:

Accuracy:

- Logistic Regression - 87.6% : generalisasi yang lumayan bagus untuk melihat data.
- Random Forest - 100% : Perfomence yang sangat bagus tapi kemungkinan besar terjadinya overfitting pada model.
- Decision Tree - 100%: Hampir sama seperti random forest namun dapat terjadinya overfitting atau tidak.

Precision:

- Logistic Regression: buat precision kelas 0 [not placed] itu tinggi ada pada 0.90, sedangkan untuk lower kelas 1 [placed] itu 0.75
- Random forest & Decision tree: dua duanya memiliki precision yang sempurna untuk semua kelas yaitu 1.0

Recall: 

- Logistic Regression: untuk recall kelas 0 itu tinggi ada pada (0.85), sedangkan untuk recall kelas 1 itu (0.65), yang dimana ada indikasi bahwa beberapa false negatives untuk kelas 1.
- Random Forest & Decision Tree: dua model tersebut memeiliki nilai recall yang sempurna untuk semua kelas yaitu 1.0 , yang dimana berati tidak memiliki false negatives.

F1-Score:

- Logistic Regression: F1-score untuk kelas 0 itu 0.87, dan untuk kelas 1 itu 0.70, yang dimana memberikan keseimbangan untuk precision dan recall.
- Random Forest & Decision Tree: dua model tersebut memeiliki nilai recall yang sempurna untuk semua kelas yaitu (1.0), memastikan keseimbangan nilai antara precision dan recall.
